src/db.c

```diff
@@ -1091,7 +1092,13 @@ void scanCallback(void *privdata, const dictEntry *de) {
     }
 
     listAddNodeTail(keys, key);
-    if (val) listAddNodeTail(keys, val);
+    if (val) {
+        if (!data->no_values) {
+            listAddNodeTail(keys, val);
+        } else {
+            sdsfree(val);
```

Comment:
It‘s better putting it within the zset scope to avoid unnecessary memory allocation and deallocation.